### PR TITLE
[WIP] 2019 a11y fixes

### PR DESCRIPTION
**Pull request by zainafzal08**
Not finished yet but this is the WIP for the addition of the a11y mode.

Currently the text app suffers from many issues that make it unusable via a screen reader, for example moving the cursor fails to make the screen reader announce the current word you are on, and on any insertion of a new line the screen reader will announce that the previous line was deleted. These arise from inherent issues in codemirror, the third party library we use to render text, after much experimentation it was concluded that the best way to provide screen reader support was to not use codemirror at all and rather implement a a11y mode which replaces the codemirror instance with a simple HTML5 textarea which by default has optimal screen reader support.

This PR implements this a11y mode.

### CODE REVIEW

File: third_party/CodeMirror/lib/codemirror.js [hunk 1]
```diff
@@ -9680,4 +9680,4 @@ CodeMirror.version = "5.39.0"
 
 return CodeMirror;
 
-})));
```
Comment (dstockwell) [inline third_party/CodeMirror/lib/codemirror.js]: revert

File: js/editor-ta.js [hunk 2]
```diff
@@ -0,0 +1,244 @@
+/**
+ * @constructor
+ * @param {DOM} elementId
```
Comment (dstockwell) [inline js/editor-ta.js]: doesn't match actual param

File: js/editor-ta.js [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+  const initFontSize = this.settings_.get('fontsize') + 'px';
+
+  this.textarea_ = document.createElement('textarea');
+  this.textarea_.setAttribute('id', 'editor-ta');
+  this.textarea_.style.fontSize = initFontSize;
+  this.textarea_.addEventListener('input', function() {
+    this.onChange();
+  }.bind(this));
+  window.addEventListener('resize', function() {
```
Comment (dstockwell) [inline js/editor-ta.js]: When is this removed?

File: js/editor-ta.js [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    const height = this.settings_.get('wordwrap') ? this.getLineHeight(line) : undefined;
+    this.lineNumbers_.appendChild(this.lineNumberElem(line, height));
+    linesRendered++;
+  }
+
+};
+
+EditorTextArea.prototype.getLineHeight = function(lineNumber) {
+  // TODO(zafzal): Speed this up, splitting the whole doc is a bottleneck
+  const line = this.textarea_.value.split('\n')[lineNumber-1];
```
Comment (dstockwell) [inline js/editor-ta.js]: the lines aren't all the same height?

File: js/editor-cm.js [hunk 5]
```diff
@@ -1,5 +1,108 @@
 var EditSession = CodeMirror.Doc;
 
+// taken from a chrome dev tools
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-cm.js]: It still improves the experience somewhat by giving partial navigation support with chromevox, i.e using the arrow keys to move around the file works a bit better. Thought there was no harm in leaving it in?

File: js/app.js [hunk 6]
```diff
@@ -81,10 +81,30 @@ TextApp.prototype.setTheme = function() {
...
+  while(editor.firstElementChild !== null) {
+    editor.firstElementChild.remove();
+  }
+}
+
 /**
  * Called when all the services have started and settings are loaded.
  */
 TextApp.prototype.onSettingsReady_ = function() {
+  if (this.settings_.get('accessibilitymode')) {
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/app.js]: Good idea! Will shift all of the editor init logic into initEditor_

File: js/editor-ta.js (hunk 3 above)
Comment (zainafzal08) [reply js/editor-ta.js]: never! I'll update the editor API to include a destruction method.

File: js/editor-ta.js [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+/**
+ * @return {Search}
+ * Return search object.
+ */
+EditorTextArea.prototype.getSearch = function() {
+  // TODO: this
+};
+
+EditorTextArea.prototype.onChange = function() {
+  this.numLines_ = this.textarea_.value.split('').filter(x => x === '\n').length + 1;
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Good idea! Will do

[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Funny enough the best solution is to use a regex replace apparently
https://jsperf.com/count-number-of-occurrences-of-a-character-in-a-string

File: js/app.js [hunk 8]
```diff
@@ -95,7 +115,7 @@ TextApp.prototype.onSettingsReady_ = function() {
   this.analytics_.setEnabled(this.settings_.get('analytics'));
```
Comment (BugsNash) [inline js/app.js]: the analytics and sindowController stuff doesn't seem editor related, should that remain in onSettingsReady?

File: js/app.js [hunk 9]
```diff
@@ -81,10 +81,30 @@ TextApp.prototype.setTheme = function() {
   this.editor_.setTheme(theme);
 };
 
+/**
+ * remove the editor so it can be reinitalised
+ */
+TextApp.prototype.removeEditor = function() {
+  const editor = $('#editor')[0];
```
Comment (BugsNash) [inline js/app.js]: nb i was trying to move away from jquery use in the hopes i would get a chance to delete it one day. fine if you want to use it though

File: js/app.js [hunk 10]
```diff
@@ -81,10 +81,30 @@ TextApp.prototype.setTheme = function() {
...
+  }
+}
+
 /**
  * Called when all the services have started and settings are loaded.
  */
 TextApp.prototype.onSettingsReady_ = function() {
+  if (this.settings_.get('accessibilitymode')) {
+    this.removeEditor();
+    this.editor_ = new EditorTextArea($('#editor')[0], this.settings_);
```
Comment (BugsNash) [inline js/app.js]: since we need to grab the editor from the DOM here I'd pass it as an arg to removeEditor rather than have to scan the DOM twice (same below)

File: js/app.js [hunk 11]
```diff
@@ -23,7 +23,7 @@ function TextApp() {
 TextApp.prototype.init = function() {
   this.settings_ = new Settings();
   this.analytics_ = new Analytics();
-  this.editor_ = new Editor($('#editor')[0], this.settings_);
+  this.editor_ = new EditorCodeMirror($('#editor')[0], this.settings_);
```
Comment (BugsNash) [inline js/app.js]: can this be moved to the initEditor method in an if/else on accessiblity mode? that way we can remove the need for the to call removeEditor on startup

File: js/app.js [hunk 12]
```diff
@@ -135,6 +155,16 @@ TextApp.prototype.onSettingsChanged_ = function(e, key, value) {
     case 'wraplines':
       this.editor_.setWrapLines(value);
       break;
+
+    case 'accessibilitymode':
+      this.removeEditor();
+      if (value) {
+        this.editor_ = new EditorTextArea($('#editor')[0], this.settings_);
```
Comment (BugsNash) [inline js/app.js]: might not be worth worrying about but i just had a thought here - is it expensive to create a new editor like this? what if the user toggles this option over and over?

File: _locales/en/messages.json [hunk 13]
```diff
@@ -67,6 +67,10 @@
     "message": "Always on top",
     "description": "The label for the checkbox in settings tab that sets the window always on top."
   },
+  "accessibilityModeSetting": {
+    "message": "Accessibility Mode",
+    "description": "The label for the checkbox in settings tab that activates accessibility mode, this mode allows for screenreaders to function correctly but disables syntax highlighting"
```
Comment (BugsNash) [inline _locales/en/messages.json]: how is it that dev tools keeps the syntax highlighting?

File: js/editor-cm.js [hunk 14]
```diff
@@ -1,5 +1,108 @@
...
+// go/codemirror-a11y-hack
+CodeMirror.inputStyles.devToolsAccessibleTextArea = class extends CodeMirror.inputStyles.textarea {
+  /**
+   * @override
+   * @param {!Object} display
+   */
+  init(display) {
+    super.init(display);
+    this.textarea.addEventListener('compositionstart', this._onCompositionStart.bind(this));
+
```
Comment (BugsNash) [inline js/editor-cm.js]: nit: extra blank line

File: js/editor-cm.js [hunk 15]
```diff
@@ -1,5 +1,108 @@
...
+  }
+
+  _onCompositionStart() {
```
Comment (BugsNash) [inline js/editor-cm.js]: better to make the function name describe what it does rather than when it's called (other places in text app are guilty of this but let's not add to it)

File: js/editor-cm.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-cm.js]: are these changes separate to the addition of the 'accessibility mode' change? if so i would pull them out into a seprate PR

File: js/editor-cm.js [hunk 16]
```diff
@@ -8,20 +111,30 @@ var EditSession = CodeMirror.Doc;
 function EditorCodeMirror(editorElement, settings) {
   this.element_ = editorElement;
   this.settings_ = settings;
+
+
```
Comment (BugsNash) [inline js/editor-cm.js]: nit: extra blank line. no need for either blank line here imo

File: js/editor-cm.js [hunk 17]
```diff
@@ -8,20 +111,30 @@ var EditSession = CodeMirror.Doc;
...
   this.cm_ = CodeMirror(
       editorElement,
       {
         'value': '',
         'autofocus': true,
         'matchBrackets': true,
+        'inputStyle': 'devToolsAccessibleTextArea',
+        // Poll interval of ~25 days, this is to prevent codemirror from
+        // deleting all of our selections
+        'pollInterval': Math.pow(2, 31) - 1,
```
Comment (BugsNash) [inline js/editor-cm.js]: does this break anything like IME input? also add to comment the units pollInterval is taking e.g., '25 days (in milliseconds).

File: js/controllers/hotkeys.js [hunk 18]
```diff
@@ -104,14 +103,3 @@ HotkeysController.prototype.onKeydown_ = function(e) {
     }
   }
 };
-
-HotkeysController.prototype.onMouseWheel_ = function(e) {
```
Comment (BugsNash) [inline js/controllers/hotkeys.js]: why is this deleted?

File: js/editor-ta.js [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+  this.mirror_.style.fontSize = fontSize + 'px';
+  this.lineMirror_.style.fontSize = fontSize + 'px';
+  this.updateLineNumbers();
+  this.updateTextArea();
+};
+
+/**
+ * @param {number} size
+ */
+EditorTextArea.prototype.setTabSize = function(size) {
```
Comment (BugsNash) [inline js/editor-ta.js]: I think it's reasonable in a v0 to disable these settings in accessibility mode

File: js/editor-cm.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-cm.js]: they are, will shift into a seperate PR :)

File: js/editor-ta.js (hunk 4 above)
Comment (zainafzal08) [reply js/editor-ta.js]: nope! if you have word wrap on the height of the line changes depending on conten :(

File: js/editor-ta.js [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+EditorTextArea.prototype.showHideLineNumbers = function(val) {
+  // TODO(zafzal): this
+};
+
+/**
+ * @param {boolean} val
+ */
+EditorTextArea.prototype.setWrapLines = function(val) {
+  // just update the text area which will check the relevent settings item
+  this.updateTextArea();
```
Comment (BugsNash) [inline js/editor-ta.js]: does it also need to do the height update that updateTextArea performs?

File: _locales/en/messages.json [hunk 21]
```diff
@@ -67,6 +67,10 @@
     "message": "Always on top",
     "description": "The label for the checkbox in settings tab that sets the window always on top."
   },
+  "accessibilityModeSetting": {
+    "message": "Accessibility Mode",
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply _locales/en/messages.json]: Currently focusing on getting the functionality working but getting UX support on how to design the experience and inform users is in the pipeline!

File: _locales/en/messages.json (hunk 13 above)
Comment (zainafzal08) [reply _locales/en/messages.json]: it doesn't have proper screen reader support

File: css/app.css [hunk 22]
```diff
@@ -461,6 +461,62 @@ header.hide-controls #window-close {
   position: relative;
 }
 
+/* These are the styles for the accessibility mode editor */
+.editor-ta-container {
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply css/app.css]: will change! Thanks :)

File: css/app.css [hunk 23]
```diff
@@ -461,6 +461,62 @@ header.hide-controls #window-close {
...
+  display: flex;
+  flex-direction: column;
+}
+
+#editor-ta-line-numbers>div {
+  width: calc(100% - 8px);
+  text-align: right;
+  padding: 0 3px 0 5px;
+  height: auto;
+  line-height: 1.2em;
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply css/app.css]: Will do!

File: css/app.css [hunk 24]
```diff
@@ -461,6 +461,62 @@ header.hide-controls #window-close {
...
+}
+
+.editor-ta-wrapper {
+  min-height: calc(100% - 8px);
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply css/app.css]: I'll put in a css variable to explain the 8px :)

File: css/app.css [hunk 25]
```diff
@@ -562,3 +618,12 @@ input[type=search]:focus {
 header.search-active ~ #editor div.CodeMirror .cm-matchhighlight {
   background-color: rgba(253, 207, 76, .5);
 }
+
+.hidden {
+  position: absolute;
+  bottom: -999999;
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply css/app.css]: I copied this css to match what codemirror does on the assumption that whatever it chose would be the most stable, in retrospect we can probably use visibility:hidden because we don't have to support IE

File: js/app.js [hunk 26]
```diff
@@ -81,10 +81,30 @@ TextApp.prototype.setTheme = function() {
   this.editor_.setTheme(theme);
 };
 
+/**
+ * remove the editor so it can be reinitalised
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/app.js]: >:O

File: js/app.js (hunk 9 above)
Comment (zainafzal08) [reply js/app.js]: i just copied and pasted this from another part of the codebase, i am personally against jquery too so happy to change this

File: js/app.js (hunk 10 above)
Comment (zainafzal08) [reply js/app.js]: good idea!

File: js/app.js (hunk 8 above)
Comment (zainafzal08) [reply js/app.js]: Very fair point, will move, thanks!

Comment (zainafzal08) [reply js/app.js]: Yep good idea

File: js/controllers/hotkeys.js (hunk 18 above)
Comment (zainafzal08) [reply js/controllers/hotkeys.js]: https://bugs.chromium.org/p/chromium/issues/detail?id=1007457&q=owner%3Ame&colspec=ID%20Pri%20M%20Stars%20ReleaseBlock%20Component%20Status%20Owner%20Summary%20OS%20Modified

Can move this to a seperate PR if you like!

File: js/editor-ta.js [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+
+  this.setTheme();
+
+  const initFontSize = this.settings_.get('fontsize') + 'px';
+
+  // The main text area the user will be interacting with
+  this.textarea_ = document.createElement('textarea');
+  this.textarea_.setAttribute('id', 'editor-ta');
+  this.textarea_.style.fontSize = initFontSize;
+  this.textarea_.addEventListener('input', function() {
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: I was avoiding arrow functions since the rest of the app was written when they didn't exist and i thought it would be a good idea to maintain the general style of the codebase. What do you think?

File: js/editor-ta.js [hunk 28]
```diff
@@ -0,0 +1,304 @@
...
+  this.container_.append(this.lineNumbers_);
+  this.container_.append(this.wrapper_);
+
+  this.containerResizeObserver_ = new ResizeObserver(function() {
+    this.onResize();
+  }.bind(this)).observe(this.container_);
+
+  this.element_.appendChild(this.container_);
+
+  // A hack which leverages a hidden div to figure out the height and width
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: not really, if you do the math with just number of lines and font size it consistently comes out too small or too big depending on font size. Most answers online seem to suggest this is the most stable way to do it

[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: height and width are important because the textarea needs to match the height of all the text + the width of the largest line so if users flip word wrap off the text area can be the correct with to let users scroll horizontally.

File: js/editor-ta.js [hunk 29]
```diff
@@ -0,0 +1,304 @@
...
+  // the text area should be since text areas can not grow with input
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: done

File: js/editor-ta.js [hunk 30]
```diff
@@ -0,0 +1,304 @@
...
+
+  // TODO: set up search
+  // TODO: setup how we are handling tabs?
+
+}
+
+/**
+ * Updates the various components of the editor when the window is resized
+ */
+EditorTextArea.prototype.onResize = function() {
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: got it!

File: js/editor-ta.js [hunk 31]
```diff
@@ -0,0 +1,304 @@
...
+  this.lineMirror_.classList.add('hidden');
+  this.lineMirror_.style.fontSize = initFontSize;
+  this.element_.appendChild(this.mirror_);
+  this.element_.appendChild(this.lineMirror_);
+
+  this.onResize();
+
+  // TODO: set up search
+  // TODO: setup how we are handling tabs?
+
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: done

File: js/editor-ta.js [hunk 32]
```diff
@@ -0,0 +1,304 @@
...
+    // accurate heights
+    this.lineMirror_.style.width = this.initWidth_ + 'px';
+  }
+
+  this.updateLineNumbers();
+  this.updateTextArea();
+}
+/**
+ * @param {number} height
+ * Updates height for both line number and container elem
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Thank you!

File: js/editor-ta.js [hunk 33]
```diff
@@ -0,0 +1,304 @@
...
+  if (this.settings_.get('wordwrap')) {
+    // TODO(zafzal): replace this with a proper remove all children
+    this.lineNumbers_.innerHTML = '';
+
+    // Update the width of the line mirror so we can extract out
+    // accurate heights
+    this.lineMirror_.style.width = this.initWidth_ + 'px';
+  }
+
+  this.updateLineNumbers();
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: ooo nice pick up! Thank you

File: js/editor-ta.js [hunk 34]
```diff
@@ -0,0 +1,304 @@
...
+ */
+EditorTextArea.prototype.lineNumberElem = function(number, height) {
+  const e = document.createElement('div');
+  e.innerText = number;
+  if (height !== undefined)
+    e.style.height = height + 'px';
+  return e;
+}
+
+EditorTextArea.prototype.updateInitDimentions = function() {
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: yep good idea!

File: js/editor-ta.js [hunk 35]
```diff
@@ -0,0 +1,304 @@
...
+  // the -/+ 8 is to account for padding
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: got it

File: js/editor-ta.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: Interesting, do you know why it the calc comes out wrong? Maybe add a code comment about that approach not working

File: js/editor-ta.js [hunk 36]
```diff
@@ -0,0 +1,304 @@
...
+/**
+ * @param {number} number
+ * @param {number} height
+ * @return {HTMLElement}
+ * Create a span for the line number gutter containing the line number
+ */
+EditorTextArea.prototype.lineNumberElem = function(number, height) {
+  const e = document.createElement('div');
+  e.innerText = number;
+  if (height !== undefined)
```
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: hmm maybe just a comment that includes info about it defaulting to 1 line height

File: js/editor-ta.js [hunk 37]
```diff
@@ -0,0 +1,304 @@
...
+  } else {
+    this.updateHeight(calculatedHeight);
+  }
+
+  if (this.settings_.get('wraplines')) {
+    this.updateWidth(this.initWidth_);
+  } else {
+    this.updateWidth(Math.max(this.initWidth_, calculatedWidth));
+  }
+
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Done

File: js/app.js [hunk 38]
```diff
@@ -81,10 +81,30 @@ TextApp.prototype.setTheme = function() {
...
 TextApp.prototype.onSettingsReady_ = function() {
+  if (this.settings_.get('accessibilitymode')) {
+    this.removeEditor();
+    this.editor_ = new EditorTextArea($('#editor')[0], this.settings_);
+  }
+
+  this.initEditor_();
+};
+
+// Loads settings into a new text area
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/app.js]: done

File: js/app.js (hunk 11 above)
Comment (zainafzal08) [reply js/app.js]: done!

File: js/app.js (hunk 12 above)
Comment (zainafzal08) [reply js/app.js]: It is expensive for sure, i'm hesitant to switch over to having both editors created once and switching them out because it introduces some complexity which i feel is excessive since it's unlikely this toggle will be switched often enough to cause a performance hit?

File: js/editor-cm.js (hunk 14 above)
Comment (zainafzal08) [reply js/editor-cm.js]: done

File: js/editor-cm.js (hunk 15 above)
Comment (zainafzal08) [reply js/editor-cm.js]: Code was killed in this PR

File: js/editor-cm.js (hunk 16 above)
Comment (zainafzal08) [reply js/editor-cm.js]: Done

File: js/editor-cm.js (hunk 17 above)
Comment (zainafzal08) [reply js/editor-cm.js]: Code was deleted from this PR

File: js/editor-ta.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: oh! In researching why this happens it turns out the issue is that variable line heights cause issues as it adds space above and below a line, but since we are explicitly setting a line height we can do the math ourselves, will try update the code :)

[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: So turns out that the height of the text is off because we were not considering line height, i  can update the calculations but there is still the issue of having to figure out the width and height of the text box given word wrap. I think this method is a good way to handle that case since writing our own code to figure out when the line wraps is a bit annoying.
That being said i will tweek the code to use calculations where possible (i.,e when word wrap is off)

File: js/editor-ta.js (hunk 36 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: done!

File: js/editor-ta.js [hunk 39]
```diff
@@ -0,0 +1,304 @@
...
+  e.innerText = number;
+  if (height !== undefined)
+    e.style.height = height + 'px';
+  return e;
+}
+
+EditorTextArea.prototype.updateInitDimentions = function() {
+  // the -/+ 8 is to account for padding
+  this.initHeight_ = this.element_.getBoundingClientRect().height - 8;
+  this.container_.style.height = (this.initHeight_ + 8) + 'px';
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Done!

File: js/editor-ta.js [hunk 40]
```diff
@@ -0,0 +1,304 @@
...
+  }
+  this.updateLineNumbers();
+  this.updateTextArea();
+  // TODO(zafzal): need to emit a docChange event to each tab can know a change occured...
+};
+
+EditorTextArea.prototype.updateTextArea = function() {
+  const calculatedHeight = this.mirror_.getBoundingClientRect().height;
+  const calculatedWidth = this.mirror_.getBoundingClientRect().width;
+  // if the height is less then the window height just snap to the window height
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Doesn't seem like it? Let me know if i've missed some configuration you can set

File: js/editor-ta.js [hunk 41]
```diff
@@ -0,0 +1,304 @@
...
+  // remove excess elements
+  while (linesRendered > this.numLines_) {
+    this.lineNumbers_.lastChild.remove();
+    linesRendered--;
+  }
+  // add missing elements
+  while(linesRendered < this.numLines_) {
+    const line = linesRendered + 1;
+    // If word wrap is on we have to additionally set height
+    const height = this.settings_.get('wordwrap') ? this.getLineHeight(line) : undefined;
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: The currently (albeit hacky) solution is that on any change when word wrap is on the entire line gutter is cleared and then repopulated so a change to a line in the middle of the doc is registered and the line height increased.

File: js/editor-ta.js [hunk 42]
```diff
@@ -0,0 +1,304 @@
...
+
+EditorTextArea.prototype.getLineHeight = function(lineNumber) {
+  // TODO(zafzal): Speed this up, splitting the whole doc is a bottleneck
+  const line = this.textarea_.value.split('\n')[lineNumber-1];
+  this.lineMirror_.innerText = line;
+  return this.lineMirror_.getBoundingClientRect().height;
+}
+
+EditorTextArea.prototype.undo = function() {
+  // This is handled by the text area defaults
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: That's a really good point, the undo stacks are separate on both so the undo stack will not be preserved. I don't think we're planning to support that at this point but it's worth pointing out on the doc, will add it

File: js/editor-ta.js (hunk 19 above)
Comment (zainafzal08) [reply js/editor-ta.js]: i believe it's actually a very easy css property you can set so no harm in keeping it

File: js/editor-ta.js (hunk 20 above)
Comment (zainafzal08) [reply js/editor-ta.js]: Yep, the height of the document changes depending on if we are wrapping lines or not

File: js/app.js (hunk 12 above)
Comment (BugsNash) [reply js/app.js]: sgtm, maybe add a code comment to that effect

File: js/editor-ta.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: sgtm

File: js/editor-ta.js (hunk 42 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: sgtm

File: js/editor-ta.js (hunk 40 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: not that i'm aware of, I'm just surprised

File: js/editor-ta.js (hunk 41 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: oh, I must have missed that code, where does that logic live?

File: js/editor-ta.js (hunk 19 above)
Comment (BugsNash) [reply js/editor-ta.js]: Ack. For any settings not yet implemented I wouldn't block this on them.

File: js/editor-ta.js (hunk 20 above)
Comment (BugsNash) [reply js/editor-ta.js]: Ack.

File: js/editor-ta.js (hunk 41 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: in updateLineNumbers the first couple lines check if wraplines is true and if so clear all the line numbers. (could have potentially been put in after you made the comment sorry!!)

[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: Ack.

File: js/editor-ta.js [hunk 43]
```diff
@@ -0,0 +1,304 @@
...
+  this.textarea_.addEventListener('input', function() {
+    this.onChange();
+  }.bind(this));
+  window.addEventListener('resize', function() {
+    this.onResize();
+  }.bind(this));
+
+  this.wrapper_ = document.createElement('div');
+  this.wrapper_.classList.add('editor-ta-wrapper');
+  this.wrapper_.appendChild(this.textarea_);
```
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: I'd like to keep it as is since codemirror does not have any UI and i really want to unify the interface that both codemirror and the text area use to minimize the chance of things going weird when you switch between the two.

File: js/editor-ta.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Will keep this as is to match the current style of the app, introducing a mix of syntax is imo not worth the benefits that a arrow function gives

File: js/controllers/hotkeys.js (hunk 18 above)
Comment (zainafzal08) [reply js/controllers/hotkeys.js]: Will move to a seperate PR.

File: js/editor-ta.js (hunk 43 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: Ack. Maybe add a comment as such?

File: js/editor-ta.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (BugsNash) [reply js/editor-ta.js]: sorry I didn't see this before! when I started working on text app I updated it to be able to work with ES6 and started using ES6 on all my changes, including arrow functions, const/let instead of ver, etc. so there is already a mix of syntax in the code base. i think all new code/changes should stick to ES6 (and ideally the old stuff updated, but not as a priority)

File: js/editor-ta.js (hunk 43 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Will do!

File: js/editor-ta.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (zainafzal08) [reply js/editor-ta.js]: Fair point, will update to arrow :)
